File: 01_File_System_Analysis/file_analyzer.go
package main

import (
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"time"
)

// FileInfo represents file metadata for forensic analysis
type FileInfo struct {
	Path         string            `json:"path"`
	Name         string            `json:"name"`
	Size         int64             `json:"size"`
	Mode         string            `json:"mode"`
	ModTime      time.Time         `json:"mod_time"`
	IsDir        bool              `json:"is_dir"`
	Hashes       map[string]string `json:"hashes"`
	MagicBytes   string            `json:"magic_bytes"`
	Extension    string            `json:"extension"`
	Permissions  string            `json:"permissions"`
}

// FileCarver performs file carving operations
type FileCarver struct {
	outputDir string
}

// NewFileCarver creates a new file carver instance
func NewFileCarver(outputDir string) *FileCarver {
	return &FileCarver{outputDir: outputDir}
}

// calculateHashes computes MD5, SHA1, and SHA256 hashes
func calculateHashes(filePath string) (map[string]string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	md5Hash := md5.New()
	sha1Hash := sha1.New()
	sha256Hash := sha256.New()

	multiWriter := io.MultiWriter(md5Hash, sha1Hash, sha256Hash)
	
	if _, err := io.Copy(multiWriter, file); err != nil {
		return nil, err
	}

	hashes := map[string]string{
		"md5":    hex.EncodeToString(md5Hash.Sum(nil)),
		"sha1":   hex.EncodeToString(sha1Hash.Sum(nil)),
		"sha256": hex.EncodeToString(sha256Hash.Sum(nil)),
	}

	return hashes, nil
}

// getMagicBytes reads the first 16 bytes of a file
func getMagicBytes(filePath string) (string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return "", err
	}
	defer file.Close()

	buffer := make([]byte, 16)
	n, err := file.Read(buffer)
	if err != nil && err != io.EOF {
		return "", err
	}

	return hex.EncodeToString(buffer[:n]), nil
}

// analyzeFile extracts comprehensive metadata from a file
func analyzeFile(filePath string) (*FileInfo, error) {
	info, err := os.Stat(filePath)
	if err != nil {
		return nil, err
	}

	var hashes map[string]string
	var magicBytes string

	if !info.IsDir() {
		hashes, _ = calculateHashes(filePath)
		magicBytes, _ = getMagicBytes(filePath)
	}

	fileInfo := &FileInfo{
		Path:        filePath,
		Name:        info.Name(),
		Size:        info.Size(),
		Mode:        info.Mode().String(),
		ModTime:     info.ModTime(),
		IsDir:       info.IsDir(),
		Hashes:      hashes,
		MagicBytes:  magicBytes,
		Extension:   filepath.Ext(filePath),
		Permissions: info.Mode().Perm().String(),
	}

	return fileInfo, nil
}

// scanDirectory recursively scans a directory and analyzes all files
func scanDirectory(dirPath string, recursive bool) ([]*FileInfo, error) {
	var results []*FileInfo

	err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Skip subdirectories if not recursive
		if !recursive && info.IsDir() && path != dirPath {
			return filepath.SkipDir
		}

		fileInfo, err := analyzeFile(path)
		if err != nil {
			log.Printf("Error analyzing %s: %v", path, err)
			return nil
		}

		results = append(results, fileInfo)
		return nil
	})

	return results, err
}

// findDeletedFiles searches for deleted file signatures in unallocated space
func (fc *FileCarver) findDeletedFiles(imagePath string) error {
	// This is a simplified implementation
	// In practice, you'd need to parse the file system structure
	file, err := os.Open(imagePath)
	if err != nil {
		return err
	}
	defer file.Close()

	buffer := make([]byte, 4096)
	fileSignatures := map[string]string{
		"ffd8ff":   "jpg",
		"89504e47": "png",
		"47494638": "gif",
		"25504446": "pdf",
		"504b0304": "zip",
	}

	offset := int64(0)
	for {
		n, err := file.Read(buffer)
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}

		hexData := hex.EncodeToString(buffer[:n])
		
		for signature, extension := range fileSignatures {
			if len(hexData) >= len(signature) {
				for i := 0; i <= len(hexData)-len(signature); i++ {
					if hexData[i:i+len(signature)] == signature {
						fmt.Printf("Found %s signature at offset %d\n", extension, offset+int64(i/2))
						// Here you would extract the file
					}
				}
			}
		}

		offset += int64(n)
	}

	return nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run file_analyzer.go <path> [recursive] [output.json]")
		return
	}

	path := os.Args[1]
	recursive := len(os.Args) > 2 && os.Args[2] == "true"
	outputFile := ""
	if len(os.Args) > 3 {
		outputFile = os.Args[3]
	}

	var results []*FileInfo
	var err error

	info, err := os.Stat(path)
	if err != nil {
		log.Fatal(err)
	}

	if info.IsDir() {
		results, err = scanDirectory(path, recursive)
	} else {
		fileInfo, err := analyzeFile(path)
		if err != nil {
			log.Fatal(err)
		}
		results = []*FileInfo{fileInfo}
	}

	if err != nil {
		log.Fatal(err)
	}

	if outputFile != "" {
		file, err := os.Create(outputFile)
		if err != nil {
			log.Fatal(err)
		}
		defer file.Close()

		encoder := json.NewEncoder(file)
		encoder.SetIndent("", "  ")
		if err := encoder.Encode(results); err != nil {
			log.Fatal(err)
		}
		fmt.Printf("Results saved to %s\n", outputFile)
	} else {
		encoder := json.NewEncoder(os.Stdout)
		encoder.SetIndent("", "  ")
		encoder.Encode(results)
	}
}

File: 01_File_System_Analysis/file_analyzer.rs
use std::fs;
use std::path::Path;
use std::io::{self, Read};
use sha2::{Sha256, Digest};
use md5;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Serialize, Deserialize, Debug)]
struct FileMetadata {
    path: String,
    name: String,
    size: u64,
    created: Option<String>,
    modified: Option<String>,
    accessed: Option<String>,
    is_directory: bool,
    permissions: String,
    file_type: String,
    md5_hash: Option<String>,
    sha256_hash: Option<String>,
    magic_bytes: Option<String>,
}

struct FileSystemAnalyzer {
    file_signatures: HashMap<Vec<u8>, &'static str>,
}

impl FileSystemAnalyzer {
    fn new() -> Self {
        let mut signatures = HashMap::new();
        signatures.insert(vec![0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A], "PNG");
        signatures.insert(vec![0xFF, 0xD8, 0xFF], "JPEG");
        signatures.insert(vec![0x47, 0x49, 0x46, 0x38, 0x37, 0x61], "GIF87a");
        signatures.insert(vec![0x47, 0x49, 0x46, 0x38, 0x39, 0x61], "GIF89a");
        signatures.insert(vec![0x25, 0x50, 0x44, 0x46], "PDF");
        signatures.insert(vec![0x50, 0x4B, 0x03, 0x04], "ZIP");
        signatures.insert(vec![0x4D, 0x5A], "PE Executable");
        signatures.insert(vec![0x7F, 0x45, 0x4C, 0x46], "ELF");
        signatures.insert(vec![0xD0, 0xCF, 0x11, 0xE0], "MS Office");
        
        FileSystemAnalyzer {
            file_signatures: signatures,
        }
    }

    fn identify_file_type(&self, file_path: &Path) -> String {
        if let Ok(mut file) = fs::File::open(file_path) {
            let mut buffer = vec![0; 32];
            if let Ok(_) = file.read(&mut buffer) {
                for (signature, file_type) in &self.file_signatures {
                    if buffer.starts_with(signature) {
                        return file_type.to_string();
                    }
                }
            }
        }
        "Unknown".to_string()
    }

    fn calculate_hashes(&self, file_path: &Path) -> (Option<String>, Option<String>) {
        if let Ok(content) = fs::read(file_path) {
            let md5_hash = format!("{:x}", md5::compute(&content));
            let sha256_hash = format!("{:x}", Sha256::digest(&content));
            (Some(md5_hash), Some(sha256_hash))
        } else {
            (None, None)
        }
    }

    fn get_magic_bytes(&self, file_path: &Path) -> Option<String> {
        if let Ok(mut file) = fs::File::open(file_path) {
            let mut buffer = vec![0; 16];
            if let Ok(bytes_read) = file.read(&mut buffer) {
                buffer.truncate(bytes_read);
                return Some(buffer.iter().map(|b| format!("{:02x}", b)).collect::<String>());
            }
        }
        None
    }

    fn analyze_file(&self, file_path: &Path) -> io::Result<FileMetadata> {
        let metadata = fs::metadata(file_path)?;
        let file_name = file_path.file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("")
            .to_string();

        let created = metadata.created().ok()
            .and_then(|t| DateTime::<Utc>::from(t).format("%Y-%m-%d %H:%M:%S UTC").to_string().into());
        
        let modified = metadata.modified().ok()
            .and_then(|t| DateTime::<Utc>::from(t).format("%Y-%m-%d %H:%M:%S UTC").to_string().into());

        let accessed = metadata.accessed().ok()
            .and_then(|t| DateTime::<Utc>::from(t).format("%Y-%m-%d %H:%M:%S UTC").to_string().into());

        let (md5_hash, sha256_hash) = if metadata.is_file() {
            self.calculate_hashes(file_path)
        } else {
            (None, None)
        };

        let file_type = if metadata.is_file() {
            self.identify_file_type(file_path)
        } else {
            "Directory".to_string()
        };

        let magic_bytes = if metadata.is_file() {
            self.get_magic_bytes(file_path)
        } else {
            None
        };

        Ok(FileMetadata {
            path: file_path.to_string_lossy().to_string(),
            name: file_name,
            size: metadata.len(),
            created,
            modified,
            accessed,
            is_directory: metadata.is_dir(),
            permissions: format!("{:o}", metadata.permissions().mode()),
            file_type,
            md5_hash,
            sha256_hash,
            magic_bytes,
        })
    }

    fn scan_directory(&self, dir_path: &Path, recursive: bool) -> io::Result<Vec<FileMetadata>> {
        let mut results = Vec::new();
        
        fn scan_recursive(
            analyzer: &FileSystemAnalyzer, 
            dir: &Path, 
            recursive: bool, 
            results: &mut Vec<FileMetadata>
        ) -> io::Result<()> {
            for entry in fs::read_dir(dir)? {
                let entry = entry?;
                let path = entry.path();
                
                match analyzer.analyze_file(&path) {
                    Ok(metadata) => results.push(metadata),
                    Err(e) => eprintln!("Error analyzing {}: {}", path.display(), e),
                }

                if recursive && path.is_dir() {
                    scan_recursive(analyzer, &path, recursive, results)?;
                }
            }
            Ok(())
        }

        scan_recursive(self, dir_path, recursive, &mut results)?;
        Ok(results)
    }
}

fn main() -> io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    
    if args.len() < 2 {
        eprintln!("Usage: {} <path> [recursive] [output.json]", args[0]);
        std::process::exit(1);
    }

    let path = Path::new(&args[1]);
    let recursive = args.get(2).map_or(false, |s| s == "true");
    let output_file = args.get(3);

    let analyzer = FileSystemAnalyzer::new();
    
    let results = if path.is_dir() {
        analyzer.scan_directory(path, recursive)?
    } else {
        vec![analyzer.analyze_file(path)?]
    };

    let json_output = serde_json::to_string_pretty(&results)
        .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;

    if let Some(output_path) = output_file {
        fs::write(output_path, json_output)?;
        println!("Results saved to {}", output_path);
    } else {
        println!("{}", json_output);
    }

    Ok(())
}
